csharp/ql/src/semmle/code/csharp/Implements.qll

```diff
@@ -232,90 +233,158 @@ private Type getArgumentOrReturnType(Method m, int i) {
+      this = TArrayTypeKind(_, _) and result = 1
       or
-      t1.(ConstructedType).getUnboundGeneric() = t2.(ConstructedType).getUnboundGeneric()
-    ) and
-    structurallyCompatibleChildren(m1, m2, t1, t2, t1.getNumberOfChildren() - 1)
-  )
-}
+      exists(UnboundGenericType ugt | this = TConstructedType(ugt) |
+        result = ugt.getNumberOfTypeParameters()
+      )
+    }
 
-// Predicate folding to force joining on `candidateTypes` first
-// to prevent `private#structurallyCompatibleTypeParameter#fbbfb`
-pragma[nomagic]
-private predicate structurallyCompatibleTypeParameter(
-  UnboundGenericMethod m1, UnboundGenericMethod m2, int i, Type t1, Type t2
-) {
-  candidateTypes(m1, m2, t1, t2) and
-  t1 = m1.getTypeParameter(i)
-}
+    string toString() {
+      this = TPointerTypeKind() and result = "*"
+      or
+      this = TNullableTypeKind() and result = "?"
+      or
+      exists(int dim, int rnk | this = TArrayTypeKind(dim, rnk) |
+        result = "[" + dim + ", " + rnk + "]"
+      )
+      or
+      exists(UnboundGenericType ugt | this = TConstructedType(ugt) |
+        result = ugt.getNameWithoutBrackets()
+      )
+    }
 
-/**
- * Holds if the `i+1` first children of types `x` and `y` are compatible.
- */
-private predicate structurallyCompatibleChildren(
-  UnboundGenericMethod m1, UnboundGenericMethod m2, Type t1, Type t2, int i
-) {
-  exists(Type t3, Type t4 |
-    i = 0 and
-    candidateChildren(t1, t2, i, t3, t4) and
-    structurallyCompatible(m1, m2, t3, t4)
-  )
-  or
-  exists(Type t3, Type t4 |
-    structurallyCompatibleChildren(m1, m2, t1, t2, i - 1) and
-    candidateChildren(t1, t2, i, t3, t4) and
-    structurallyCompatible(m1, m2, t3, t4)
-  )
-}
+    Location getLocation() { result instanceof EmptyLocation }
+  }
 
-// manual magic predicate used to enumerate types relevant for structural comparison
-private predicate candidateTypes(UnboundGenericMethod m1, UnboundGenericMethod m2, Type t1, Type t2) {
-  m2 = getAnInterfaceMethodCandidate(m1) and
-  (
-    exists(int i |
-      t1 = getArgumentOrReturnType(m1, i) and
-      t2 = getArgumentOrReturnType(m2, i)
-    )
+  /** Gets the type kind for type `t`, if any. */
+  CompoundTypeKind getTypeKind(Type t) {
+    result = TPointerTypeKind() and t instanceof PointerType
+    or
+    result = TNullableTypeKind() and t instanceof NullableType
+    or
+    t = any(ArrayType at | result = TArrayTypeKind(at.getDimension(), at.getRank()))
     or
-    exists(Type t3, Type t4, int j |
-      candidateTypes(m1, m2, t3, t4) and
-      t1 = t3.getChild(j) and
-      t2 = t4.getChild(j)
+    result = TConstructedType(t.(ConstructedType).getUnboundGeneric())
+  }
+
+  private class MethodTypeParameter extends TypeParameter {
+    MethodTypeParameter() { this = any(UnboundGenericMethod ugm).getATypeParameter() }
+  }
+
+  private class LeafType extends Type {
+    LeafType() {
+      not exists(this.getAChild()) and
+      not this instanceof MethodTypeParameter
```

Comment:
Why are `MethodTypeParameter` and `TypeTypeParameter` handled differently? You could have a constructed type having a type parameter that is a type parameter (from another unbound generic type).

Reply:
Again, see my comment above.